services/housekeeper/housekeeper.go

```diff
@@ -252,3 +346,49 @@ func (hk *Housekeeper) updateProposerDuties(headSlot uint64) {
 	sort.Strings(_duties)
 	log.WithField("numDuties", len(_duties)).Infof("proposer duties updated: %s", strings.Join(_duties, ", "))
 }
+
+func (hk *Housekeeper) getProposerDuties(epoch uint64) *beaconclient.ProposerDutiesResponse {
+	// return the first successful beacon node response
+	var result *beaconclient.ProposerDutiesResponse
+	clients := hk.getBeaconClientsByLastResponse()
+
+	var mu sync.Mutex
+	for i, client := range clients {
+		log := hk.log.WithField("uri", client.GetURI())
+		log.Debug("fetching proposer duties")
+
+		duties, err := client.GetProposerDuties(epoch)
+		if err != nil {
+			hk.log.WithError(err).Error("failed to get proposer duties")
+			continue
+		}
+
+		// Received successful response. Set this index as last successful beacon node
+		result = duties
+
+		mu.Lock()
+		hk.lastHealthlyBeaconNodeIndex = i
+		mu.Unlock()
+
+		break
+	}
+
+	return result
+}
+
+func (hk *Housekeeper) getBeaconClientsByLastResponse() []beaconclient.BeaconNodeClient {
+	// get beacon node with last successful response
+	var mu sync.Mutex
```

Comment:
remove the mutex?

Reply:
then you'd need a single mutex used across these functions (eg. a prop on the housekeeper struct). but you are using a local unique mutex that's unaffected by other functions.